### ISSUE TITLE
Use pattern matching for instanceof where appropriate

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#28181]
```java
if (obj instanceof String) {
    String s = (String) obj;
    // use s
}
```

becomes

```java
if (obj instanceof String s) {
    // use s
}
```

**Comment by wilkinsona** [comment]
Not at the moment. Thanks for the offer, though. We haven't started work on 3.0 yet and there are a number of things that'll have to be tackled before we can look at this one.

**Comment by wilkinsona** [comment]
Closing in favor of #31475.
